scalalib/src/mill/scalalib/cosmo/CosmoModule.scala

```diff
@@ -0,0 +1,155 @@
+package mill.scalalib.cosmo
+
+import mill.{T, Task}
+import mill.api.{PathRef, Result}
+import mill.scalalib._
+
+/**
+ * Provides a [[CosmoModule.cosmoAssembly task]] to build a cross-platfrom executable assembly jar
+ * by prepending an [[https://justine.lol/ape.html Actually Portable Executable (APE)]]
+ * launcher binary compiled with the [[https://justine.lol/cosmopolitan/index.html Cosmopolitan Libc]]
+ */
+trait CosmoModule extends mill.Module with AssemblyModule {
+  def finalMainClass: T[String]
+
+  def forkArgs(argv0: String): Task[Seq[String]] = Task.Anon { Seq[String]() }
+
+  def cosmoccVersion: T[String] = Task { "" }
+
+  def cosmocc: T[PathRef] = Task(persistent = true) {
+    val version = if (cosmoccVersion().isEmpty) "" else s"-${cosmoccVersion()}"
+    val versionedCosmocc = s"cosmocc${version}"
+    val zip = Task.dest / s"${versionedCosmocc}.zip"
+    val dest = Task.dest / versionedCosmocc / "bin" / "cosmocc"
+
+    if (!os.exists(dest)) {
+      os.write.over(
+        zip,
+        requests.get.stream(s"https://cosmo.zip/pub/cosmocc/${versionedCosmocc}.zip")
+      )
+      os.remove.all(Task.dest / versionedCosmocc)
+      os.call(("unzip", zip, "-d", Task.dest / versionedCosmocc))
+    }
+
+    PathRef(dest)
+  }
+
+  def cosmoLauncherScript: T[String] = Task {
+    val start = 1
+    val size0 = start + forkArgs().length
+    val addForkArgs = forkArgs()
+      .zip(start until size0)
+      .map { (arg, i) =>
+        s"""all_argv[${i}] = (char*)malloc((strlen("${arg}") + 1) * sizeof(char));
+           |strcpy(all_argv[${i}], "${arg}");
+        """.stripMargin
+      }.mkString("\n")
+
+    val args = forkArgs("%s")()
+    val size = size0 + args.length
+    val addForkArgsArgv0 = args
+      .zip(size0 until size)
+      .map { (arg, i) =>
+        s"""char* s${i};
+           |asprintf(&s${i}, "${arg}", argv[0]);
+           |all_argv[${i}] = (char*)malloc((strlen(s${i}) + 1) * sizeof(char));
+           |strcpy(all_argv[${i}], s${i});
+        """.stripMargin
+      }.mkString("\n");
+
+    val preArgvSize = size + 3;
+
+    s"""#include <stdlib.h>
```

Comment:
Let's move this to a `launcher.c` file in `scalalib/resources`; It might not be exactly valid C due to the template interpolations, but at least people editing it in IDEs would get some syntax highlighting to help